File: backend/src/routes/attack_chains.rs
//! Attack chain API routes.
//!
//! Provides endpoints for viewing per-application correlation group
//! summaries and detailed attack chain breakdowns.

use axum::{
    extract::{Path, Query, State},
    Json,
};
use uuid::Uuid;

use crate::errors::{ApiResponse, AppError};
use crate::middleware::auth::CurrentUser;
use crate::models::pagination::{PagedResult, Pagination};
use crate::services::attack_chains::{
    self, AppAttackChainDetail, AppAttackChainSummary, AttackChainFilters,
};
use crate::AppState;

/// GET /api/v1/attack-chains -- list applications with attack chain summaries.
pub async fn list(
    State(state): State<AppState>,
    _user: CurrentUser,
    Query(pagination): Query<Pagination>,
    Query(filters): Query<AttackChainFilters>,
) -> Result<Json<ApiResponse<PagedResult<AppAttackChainSummary>>>, AppError> {
    let result = attack_chains::list_summaries(&state.db, &filters, &pagination).await?;
    Ok(ApiResponse::success(result))
}

/// GET /api/v1/attack-chains/:app_id -- get attack chains for one application.
pub async fn get_by_app(
    State(state): State<AppState>,
    _user: CurrentUser,
    Path(app_id): Path<Uuid>,
    Query(filters): Query<AttackChainFilters>,
) -> Result<Json<ApiResponse<AppAttackChainDetail>>, AppError> {
    let detail = attack_chains::get_by_app(&state.db, app_id, &filters).await?;
    Ok(ApiResponse::success(detail))
}

File: frontend/src/api/attack-chains.ts
import { apiGet } from './client'
import type {
  AppAttackChainSummary,
  AppAttackChainDetail,
  AttackChainFilters,
} from '@/types/attack-chains'
import type { PagedResult } from '@/types/finding'

/** GET /attack-chains — list attack chain summaries for all applications. */
export function listAttackChains(
  filters: AttackChainFilters = {},
  page = 1,
  perPage = 25,
): Promise<PagedResult<AppAttackChainSummary>> {
  const params: Record<string, string> = {
    page: String(page),
    per_page: String(perPage),
  }
  if (filters.branch) params.branch = filters.branch

  return apiGet<PagedResult<AppAttackChainSummary>>('/attack-chains', params)
}

/** GET /attack-chains/:app_id — get detailed attack chains for a specific application. */
export function getAttackChainsByApp(
  appId: string,
  filters: AttackChainFilters = {},
): Promise<AppAttackChainDetail> {
  const params: Record<string, string> = {}
  if (filters.branch) params.branch = filters.branch

  const hasParams = Object.keys(params).length > 0
  return apiGet<AppAttackChainDetail>(
    `/attack-chains/${appId}`,
    hasParams ? params : undefined,
  )
}
